src/election.ts

```diff
@@ -0,0 +1,167 @@
+import BigNumber from 'bignumber.js';
+import { Lease } from './lease';
+import { Namespace } from './namespace';
+
+/**
+ * Implmentation of etcd election.
+ * @see https://github.com/coreos/etcd/blob/master/clientv3/concurrency/election.go
+ *
+ * @example
+ * const client = new Etcd3()
+ * const election = new Election(client, 'singleton_service')
+ * const id = BigNumber.random().toString()
+ *
+ * // waiting for election ready
+ * await election.ready()
+ *
+ * // process will hang here until elected
+ * await election.campaign(id)
+ */
+export class Election {
+  public static readonly prefix = 'election';
+  public static readonly notLeaderError = new Error('election: not leader');
+  public static readonly noLeaderError = new Error('election: no leader');
+  public static readonly notReadyError = new Error('election: no ready');
+
+  public readonly namespace: Namespace;
+  public readonly lease: Lease;
+
+  private _leaseId = '';
+  private _leaderKey = '';
+  private _leaderRevision = '';
+  private _isCampaigning = false;
+
+  public get leaseId(): string { return this._leaseId; }
+  public get leaderKey(): string { return this._leaderKey; }
+  public get leaderRevision(): string { return this._leaderRevision; }
+  public get isReady(): boolean { return this._leaseId.length > 0; }
+  public get isCampaigning(): boolean { return this._isCampaigning; }
+
+  constructor(namespace: Namespace,
+              public readonly name: string,
+              public readonly ttl: number = 60) {
+    this.namespace = namespace.namespace(this.getPrefix());
+    this.lease = this.namespace.lease(ttl);
+  }
+
+  public async ready() {
+    const leaseId = await this.lease.grant();
+
+    if (!this.isReady) {
+      this._leaseId = leaseId;
+    }
+  }
+
+  public async campaign(value: string) {
+    this.throwIfNotReady();
```

Comment:
How about awaiting the lease grant instead? I.e. why do callers necessarily need to know to call `await ready()`?